elasticdl/python/data/odps_io.py

```diff
@@ -102,146 +103,106 @@ def __init__(
-            raise ValueError(
-                "Total worker number is 0. Please check if table has data."
-            )
-
-        with Executor(max_workers=self._num_processes) as executor:
-
-            futures = Queue()
-            # Initialize concurrently running processes according
-            # to `num_processes`
-            for i in range(self._num_processes):
-                range_start = worker_items_with_epoch[i]
-                range_end = min(range_start + large_batch_size, table_size)
-                future = executor.submit(
-                    self.read_batch, range_start, range_end, columns
-                )
-                futures.put(future)
-
-            worker_items_index = self._num_processes
-
-            while not futures.empty():
-                if worker_items_index < len(worker_items_with_epoch):
-                    range_start = worker_items_with_epoch[worker_items_index]
-                    range_end = min(range_start + large_batch_size, table_size)
-                    future = executor.submit(
-                        self.read_batch, range_start, range_end, columns
-                    )
-                    futures.put(future)
-                    worker_items_index = worker_items_index + 1
-
-                head_future = futures.get()
-                records = head_future.result()
-                for i in range(0, len(records), batch_size):
-                    yield records[i : i + batch_size]  # noqa: E203
-
-    def read_batch(self, start, end, columns=None, max_retries=3):
+        The parallel reader launches multiple worker processes to read
+        records from an ODPS table and applies `transform_fn` to each record.
+        If `transform_fn` is not set, the transform stage will be skipped.
+
+        Worker process:
+        1. get a shard from index queue, the shard is a pair (start, count)
+            of the ODPS table
+        2. reads the records from the ODPS table
+        3. apply `transform_fn` to each record
+        4. put records to the result queue
+
+        Main process:
+        1. call `reset` to create a number of shards given a input shard
+        2. put shard to index queue of workers in round-robin way
+        3. call `get_records`  to get transformed data from result queue
+        4. call `stop` to stop the workers
         """
-        Read ODPS table in chosen row range [ `start`, `end` ) with the
-        specified columns `columns`.
-
-        Args:
-            start: The row index to start reading.
-            end: The row index to end reading.
-            columns: The list of column to read.
-            max_retries : The maximum number of retries in case of exceptions.
+        self._result_queue = Queue()
+        self._index_queues = []
+        self._workers = []
+
+        self._shards = []
+        self._shard_idx = 0
+        self._worker_idx = 0
+
+        for i in range(self._num_processes):
+            index_queue = Queue()
```

Comment:
It is better we give a predefined size queue size to avoid OOM.

Reply:
This is the queue for index of a task, which stores several (start, end) pairs. So it occupies very little memory.